src/Plugin/Deriver/ContactsEntityBlockDeriver.php

```diff
@@ -53,28 +53,53 @@ public static function create(ContainerInterface $container, $base_plugin_id) {
       if ($entity_type_id == 'user' || $has_owner) {
         $has_forms = $entity_type->hasFormClasses();
 
-        // Basic definition.
-        $this->derivatives[$entity_type_id] = $base_plugin_definition;
-        $this->derivatives[$entity_type_id]['admin_label'] = $this->t('Contacts entity form (@label)', ['@label' => $entity_type->getLabel()]);
-
-        // The entity is required for the user or types that don't have forms.
-        $this->derivatives[$entity_type_id]['context']['entity'] = new ContextDefinition('entity:' . $entity_type_id, $entity_type->getLabel(), $entity_type_id == 'user' || !$has_forms);
-        // If this can have an owner, add the context so we can do create forms.
-        if ($has_owner && $has_forms) {
-          $this->derivatives[$entity_type_id]['context']['user'] = new ContextDefinition('entity:user', $this->t('User'), FALSE);
+        // Expand out a derivative per entity bundle.
+        if ($bundle_entity_type = $entity_type->getBundleEntityType()) {
+          $bundle_types = $this->entityTypeManager->getStorage($bundle_entity_type)->loadMultiple();
+          $bundles = array_keys($bundle_types);
         }
+        else {
+          $bundles = [$entity_type_id];
+        }
+
+        foreach ($bundles as $bundle) {
+          $derivative_key = "{$entity_type_id}-{$bundle}";
 
-        // Add a few other pieces of info.
-        $this->derivatives[$entity_type_id]['_entity_type_id'] = $entity_type_id;
-        $this->derivatives[$entity_type_id]['_has_forms'] = $has_forms;
-        $this->derivatives[$entity_type_id]['_allow_create'] = $has_owner && $has_forms;
-        $this->derivatives[$entity_type_id]['_bundle_key'] = $entity_type->getKey('bundle');
+          // Basic definition.
+          $this->derivatives[$derivative_key] = $base_plugin_definition;
+          $this->derivatives[$derivative_key]['admin_label'] = $this->t('Contacts entity form (@label)', ['@label' => $entity_type->getLabel()]);
+
+          // The entity is required for the user or types that don't have forms.
+          $this->derivatives[$derivative_key]['context']['entity'] = new ContextDefinition('entity:' . $entity_type_id, $entity_type->getLabel(), $entity_type_id == 'user' || !$has_forms);
+          // If this has an owner, add the context so we can do create forms.
+          if ($has_owner && $has_forms) {
+            $this->derivatives[$derivative_key]['context']['user'] = new ContextDefinition('entity:user', $this->t('User'), FALSE);
+          }
+
+          // Add a few other pieces of info.
+          $this->derivatives[$derivative_key]['_entity_type_id'] = $entity_type_id;
+          $this->derivatives[$derivative_key]['_has_forms'] = $has_forms;
+          $this->derivatives[$derivative_key]['_allow_create'] = $has_owner && $has_forms;
+          $this->derivatives[$derivative_key]['_bundle_key'] = $entity_type->getKey('bundle');
```

Comment:
Do we need to pass in which bundle derivative this is? At the moment there's no validation that the given entity is of the correct bundle, as the context is entity type rather than bundle specific... This is perhaps even an argument for why this may not be advantageous?

Reply:
Ok lets chat